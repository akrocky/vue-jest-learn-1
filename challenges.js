export default [
  {
    name: 'Plain Text',
    sectionNum: 1,
    fullText: 'oatmeal',
    textToMatch: 'oatmeal',
    info: 'Type exact characters to match.'
  },
  {
    name: 'Phrases',
    sectionNum: 1,
    fullText: 'Avocado toast is awesome!',
    textToMatch: 'Avocado toast is awesome!',
    info: `Try us a magic character like \`.\` (period).
      That matches any character.
      To match every character add a plus sign: \`.+\`
    `
  }
]
