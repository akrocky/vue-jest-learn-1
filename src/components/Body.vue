<template>
  <v-row>
    <v-col sm="10" md="8">
      <v-row cols="12">
        <v-col>
          <h1>
            {{ $store.getters.currentChallenge.name }}
          </h1>
        </v-col>
      </v-row>

      <v-row>
        <div class="challenge-tray">
          <div class="challenge-tray-inner">
            <v-row cols="12">
              <v-col class="my-5">
                <h2
                  v-for="(text, i) in $store.getters.currentChallenge.fullText"
                  :key="text + i"
                  v-html="$options.filters.highlightFilter(
                    text,
                    userRegex,
                    userFlags,
                    $style
                  )"
                  style="white-space: pre;"
                >
                </h2>
              </v-col>
            </v-row>
          </div>
        </div>
      </v-row>

      <v-row>
        <v-col cols="9">
          <v-text-field
            outlined
            :error="regexError"
            :error-messages="errorMessage"
            v-model="userRegex"
            placeholder="Enter Regex"
            @keypress.enter.stop="nextChallenge"
            @keyup="checkValidRegex"
          >
            <template v-slot:prepend>
              /
            </template>
            <template v-slot:append-outer>
              /
            </template>
          </v-text-field>
        </v-col>

        <v-col cols="3">
          <v-text-field
            :disabled="!$store.getters.currentChallenge.flags"
            outlined
            v-model="userFlags"
            placeholder="Enter Flags"
            @keypress.enter="nextChallenge"
          />
        </v-col>
      </v-row>

      <v-row>
        <v-col cols="12">
          <v-card
            outlined
            tile
          >
            <v-card-text>
              <v-icon color="info">mdi-alert-circle-outline</v-icon>
              {{ $store.getters.currentChallenge.info }}
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
    </v-col>
    <v-snackbar
      v-model="snackbar.show"
      :color="snackbar.color"
      top
      right
    >
      {{ snackbar.message }}
    </v-snackbar>
  </v-row>
</template>

<script>
import _ from 'lodash'
import doesMatch from '@/utils/doesMatch'
import doesNotMatch from '@/utils/doesNotMatch'

export default {
  data () {
    return {
      userRegex: '',
      userFlags: '',
      regexError: false,
      errorMessage: '',
      flagsError: false,
      snackbar: {
        show: false,
        color: 'error',
        message: 'Incorrect answer. Try again.'
      }
    }
  },
  filters: {
    highlightFilter (value, userRegex, userFlags, $style) {
      try {
        const regex = new RegExp(userRegex, userFlags)
        const newValue = value.replace(
          regex, (text) => `<span class="${$style.highlight}">${text}</span>`
        )
        return newValue
      } catch {
        return value
      }
    }
  },
  methods: {
    checkValidRegex () {
      try {
        new RegExp(this.userRegex, this.userFlags)
        this.regexError = false
        this.errorMessage = ''
      } catch (err) {
        this.regexError = true
        this.errorMessage = 'Invalid Regex'
      }
    },
    nextChallenge () {
      const { fullText, pattern, flags, forbiddenPatterns } = this.$store.getters.currentChallenge

      const forbidden = forbiddenPatterns.some((value) => {
        return this.userRegex.includes(value)
      })
      if (forbidden) {
        this.showSnackbar('warning')
        return false
      }

      const currentIndex = this.$store.state.currentIndex
      const solutionRegex = new RegExp(pattern, flags)
      const userRegex = new RegExp(this.userRegex, this.userFlags)

      let validFlags = false
      let isFullTextMatched = false
      let isExtraTextMatched = false

      if (!flags || (this.userFlags && this.userFlags.includes(flags))) {
        validFlags = true
        const args = {
          fullText,
          solutionRegex,
          userRegex
        }
        isFullTextMatched = doesMatch(args)
        isExtraTextMatched = doesNotMatch(args)
      }

      if (validFlags && isFullTextMatched && isExtraTextMatched) {
        this.userRegex = ''
        this.userFlags = ''
        this.showSnackbar('success')

        this.$store.dispatch('navigateToChallenge', currentIndex + 1)
      } else {
        this.showSnackbar('error')
      }
    },
    showSnackbar (type) {
      this.snackbar.color = type
      this.snackbar.show = true

      switch (type) {
        case 'error':
          this.snackbar.message = 'Incorrect answer. Try again.'
          break
        case 'warning':
          this.snackbar.message = 'You are using a Regex Pattern that is outside of the scope of this challenge. Please refactor.'
          break
        case 'success':
          this.snackbar.message = 'Great job! Challenge completed.'
          break
      }
    }
  }
}
</script>

<style lang="sass" module>
.highlight
  background-color: yellow
</style>

<style lang="sass">
.v-text-field fieldset, .v-text-field .v-input__control, .v-text-field div.v-input__slot
  border-radius: 0
h1
  font-family: 'Baloo Thambi 2', sans-serif
.challenge-tray
  background: center center url('../assets/orange-tray2.png') no-repeat
  background-size: 100% 80%
.challenge-tray-inner
  margin: 70px 50px
</style>
