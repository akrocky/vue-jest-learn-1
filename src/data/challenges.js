export default [
  {
    name: 'Match Text',
    fullText: ['oatmeal'],
    toMatch: ['oatmeal'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'Regex will match characters from right to left. Type exact characters to match.'
  },
  {
    name: 'Match Lowercase',
    fullText: ['toast'],
    toMatch: ['toast'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'Use a capture group for lowercase letters: [a-z]+.'
  },
  {
    name: 'Match Uppercase',
    fullText: ['TOAST'],
    toMatch: ['TOAST'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'Use a capture group for uppercase letters: [A-Z]+.'
  },
  {
    name: 'Match Mixed-cases',
    fullText: ['Avocado'],
    toMatch: ['Avocado'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'Use a capture group for uppercase letters: [A-Za-z]+.'
  },
  {
    name: 'Match Numbers',
    fullText: ['12345'],
    toMatch: ['12345'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'Use a capture group for numbers: [0-9]+.'
  },
  {
    name: 'Phrases',
    fullText: ['Avocado toast is awesome'],
    toMatch: ['Avocado toast is awesome'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'You can match the spaces in a string as well by just adding the space character in your regex. Match with spaces: [A-Z a-z].'
  },
  {
    name: 'Match Word in Text',
    fullText: ['Vermont Maple Syrup'],
    toMatch: ['Maple'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'You can use regex to match one part of a string. In this case, use the techniques you have learned in previous lessons to match the word, "Maple".'
  },
  {
    name: 'Match First Word in Text',
    fullText: ['Vermont Maple Syrup', 'Maple Syrup from Vermont', 'Vermont has the Best Syrup', 'Who likes Vermont Syrup?'],
    forbiddenPatterns: ['.'],
    toMatch: ['Vermont'],
    flags: 'g',
    info: 'You can use regex to match only what\'s at the beginning of a string. Use the \'^\' at the start of your regex to match the word, "Vermont".'
  },
  {
    name: 'Match Last Word in Text',
    fullText: ['Vermont Maple Syrup'],
    toMatch: ['Syrup'],
    forbiddenPatterns: ['.'],
    flags: '',
    info: 'You can use regex to match one part of a string. In this case, use the techniques you have learned in previous lessons to match the word, "Maple".'
  },
  {
    name: 'Match Subdomains',
    fullText: ['http://example.com\nhttps://mysite.example.com\nhttps://www.example.com\nhttp://blog.example.com'],
    toMatch: ['https://mysite.example.com', 'http://blog.example.com'],
    forbiddenPatterns: ['.'],
    flags: 'g',
    info: 'Match only the subdomains and not the domains without a subdomain. \'www\' counts as part of the domain and not a subdomain. Your regex should match both http and https.'
  },
  {
    name: 'Use Global Flag',
    fullText: ['This plate is larger than that plate.'],
    toMatch: ['plate'],
    forbiddenPatterns: ['.'],
    flags: 'g',
    info: 'You can use regex to match a string multiple times...'
  },
  {
    name: 'Password Requirements',
    fullText: ['Abc$123\n123Abc_zyx\nDunkin_donuts\nBreakfast_is_gre@t'],
    toMatch: ['123Abc_zyx', 'Breakfast_is_gre@t'],
    forbiddenPatterns: ['.'],
    flags: 'g',
    info: 'What\'s an important part of any healthy breakfast? Good password requirements! Use regex to select only the passwords that match the following criteria: 1 Uppercase, 1 Lowercase, 1 Number, 1 Special Character, At Least 10 Characters Long.'
  }
]
