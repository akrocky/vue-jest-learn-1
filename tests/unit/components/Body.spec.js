import { mount } from '@vue/test-utils'

import Body from '@/components/Body.vue'
import challenges from '@/data/challenges'

describe('Body.vue', () => {
  let $store = {}
  beforeEach(() => {
    $store = {
      state: {},
      getters: {
        currentChallenge: challenges[0]
      }
    }
  })

  test('snapshot', () => {
    const wrapper = mount(Body, {
      mocks: {
        $store
      }
    })
    expect(wrapper).toMatchSnapshot()
  })

  test('challenges advance', async () => {
    const wrapper = mount(Body, {
      mocks: {
        $store
      }
    })

    // enter correct answer into input
    const input = wrapper.find('[data-test="regex"]')
    await input.setValue(challenges[0].pattern)
    // press enter
    input.trigger('keypress.enter.stop')
    // check that nextChallenge has been called
    jest.spyOn(wrapper.vm, 'nextChallenge')

    expect(wrapper.vm.nextChallenge)
      .toHaveBeenCalled()
  })
})
