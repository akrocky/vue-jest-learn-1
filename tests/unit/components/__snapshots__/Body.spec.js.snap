// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body.vue snapshot 1`] = `
<div class="row">
  <div class="col-sm-10 col-md-8 col">
    <div cols="12" class="row">
      <div class="col">
        <h1>
          Match Text
        </h1>
      </div>
    </div>
    <div class="row">
      <div class="challenge-tray">
        <div class="challenge-tray-inner">
          <div cols="12" class="row">
            <div class="my-5 col">
              <h2 style="white-space: pre;"><span class="highlight"></span>oatmeal</h2>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="row">
      <div class="col col-9">
        <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
          <div class="v-input__prepend-outer">
            /
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><input data-test="regex" id="input-2" placeholder="Enter Regex" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
          <div class="v-input__append-outer">
            /
          </div>
        </div>
      </div>
      <div class="col col-3">
        <div class="v-input v-input--is-disabled theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><input disabled="disabled" id="input-5" placeholder="Enter Flags" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="row">
      <div class="col col-12">
        <div class="v-card v-card--outlined v-sheet v-sheet--tile theme--light">
          <div class="v-card__text"><i aria-hidden="true" class="v-icon notranslate mdi mdi-alert-circle-outline theme--light info--text"></i>
            Regex will match characters from right to left. Type exact characters to match.
          </div>
        </div>
      </div>
    </div>
  </div>
  <transition-stub name="v-snack-transition"></transition-stub>
</div>
`;
